Step3

manami-bunbun opened the pull request:
## What
<!--- Write the change being made with this pull request --->

## CHECKS :warning:

Please make sure you are aware of the following.

- [x] **The changes in this PR doesn't have private information

manami-bunbun committed:
add document

manami-bunbun committed:
done step3-4

manami-bunbun committed:
wkaranai2-4-DB

manami-bunbun commented:
Databaseへの接続とitem.dbに入っているスキーマの活用法がわかりません。

item.dbを実行させて、mercari.sqlite3を作り、各エンドポイントでmercari.sqlite3を操作するイメージで書いているのですが、合っているかわかりません。

manami-bunbun commented on go/app/main.go:
```diff
@@ -52,6 +104,13 @@ func getImg(c echo.Context) error {
 
 func main() {
+	// db接続
+	db, err := sql.Open("sqlite3", "db/mercari.sqlite3")
```
ここでデータベースに繋ぎたいのですが、item.dbとmercari.sqlite3の活用のタイミングがそれぞれわかっておりません。
ここで`item.db`にしても`mercari.sqlite3`にしてもエラーが出ます。

manami-bunbun replied:
#3

KengoA replied:
hint: check the directory structure again and see where mercari.sqlite3 is! is `db` inside of `mercari-build-training-2022/go` directory?

manami-bunbun commented on go/app/main.go:
```diff
@@ -1,38 +1,89 @@
 	message := fmt.Sprintf("item received: %s", name)
 	res := Response{Message: message}
 
+	// data := Items{Items: Data{name, category}}
```
このコメントアウトは、jsonの部分で使ったものをなんとかdbに書き込むように変えようとしています

manami-bunbun replied:
#3

manami-bunbun committed:
step3-4 done in python(maybe)

manami-bunbun committed:
some changes

manami-bunbun commented:
I could create `item.db` and changed endpoints for step3-4,
but the output is `{"items":["[(111, 'namae', 'tops'), (62, 'jacket', 'fashion'), (8, 'jacket', 'fashion'), (68, 'hogehoge', 'bags')]"]}`
I'm not sure how to get db like json

manami-bunbun commented on python/database.py:
```diff
@@ -0,0 +1,28 @@
+    
+def add_item(id,name,category):
+    c.execute("INSERT INTO items(id,name,category) VALUES(?,?,?)", (id,name,category))
+    
```
will be deleted later

manami-bunbun commented on python/main.py:
```diff
@@ -4,6 +4,21 @@
+
+#テーブルを作成
+# c.execute("CREATE TABLE `items` (`id` int, `name` string,`category` string);")
+
```
ここでスキーマ(item"s".db)を活用する気がする......?

manami-bunbun committed:
minor change

manami-bunbun commented on db/items.db:
```diff
@@ -0,0 +1,5 @@
+    `id` int,
+    `name` string,
+    `category` string
+);
```
このファイルの使い所がわからないです...docker?

KengoA replied:
ここちょっと問題の書き方が分かりづらいのですが、DB自体は`mercari.sqlite3`なので`items.db`はスキーマを保存しておくだけ(特に他から利用は一旦しない)みたいなイメージです！この辺りネーミング難しいのですが、僕はDB自体を`mercari-training-build.db`, スキーマを`schema.sql`にして保存しています

manami-bunbun replied:
なるほど、疑問が解決してとてもスッキリしました。詰まってるところを汲み取ってくださり、ありがとうございます！

manami-bunbun commented on python/main.py:
```diff
@@ -18,13 +33,34 @@
+def show_item():
+    content = show_all()
+    # return '{"items":' + content + '}'
+    return {"items": {content}}
```
ここが問題の出力`{"items":["[(111, 'namae', 'tops'), (62, 'jacket', 'fashion'), (8, 'jacket', 'fashion'), (68, 'hogehoge', 'bags')]"]}`です

manami-bunbun committed:
to step3-5

manami-bunbun commented on python/main.py:
```diff
@@ -18,15 +38,62 @@
+    content = db_toList(items)
+    conn.close()
+    return {"items": f"{json.dumps(content)}"}
+
```
**Step3-4: Write into a database**
Expected response: 
`{"items": [{"name": "jacket", "category": "fashion"}, ...]}`

Current response: 
`{"items":"[{\"name\": \"hogehoge\", \"category\": \"bags\"}, {\"name\": \"lisa\", \"category\": \"book\"}]"}`

[備考]
- `return {"items": content}`にしても結果は同じ
- type (content)はlist

e-yang08 replied:
ダブルクォーテーションにバックスラッシュがついてしまうのは、double encodingしているのが原因かもしれないです。※ json.dumpsがなぜdouble encodingしてしまうかについては[こちら](https://stackoverflow.com/questions/25242262/dump-to-json-adds-additional-double-quotes-and-escaping-of-quotes)にまとまっています。
`return {"items": content}`にしても結果が同じなのは、`db_toList() method`でもともと`json.dumps(items)‘していたのが原因かもと推測しているのですが、詳細はまだわかっていないです。

tkat0 replied:
もしかすると、FastAPIのドキュメントが参考になるかもしれません。
JSONを正しく返すために、このライブラリではどのようにするのが良いのか??

https://fastapi.tiangolo.com/ja/advanced/custom-response/
https://fastapi.tiangolo.com/ja/advanced/response-directly/

manami-bunbun replied:
お二方とも、ありがとうございます！

tkatoさんに頂いた
https://fastapi.tiangolo.com/ja/advanced/custom-response/
を参照したところ、以下のようにレスポンスが返ってきて、解決しました！ありがとうございます！

`{"items":[{"name":"hogehoge","category":"bags"},{"name":"lisa","category":"book"},{"name":"jacket","category":"fashion"}]}`

Youqiさんも、正直json.dumpsもよくわからないまま試していたのでリファレンスいただいて勉強になりました、ありがとうございました！

manami-bunbun commented on python/main.py:
```diff
@@ -18,15 +38,62 @@
+    content = db_toList(items)
+    conn.close()
+    return {"items": f"{json.dumps(content)}"}
+
```
**Step3-5:  Search for an item**
Expected response:
`{"items": [{"name": "jacket", "category": "fashion"}, ...]}`
Current response:
` {“items":"[]"}`
[備考]
- 3-4の部分でレスポンスが帰ってくるので、item.dbは問題があることは考えにくい
- 'keyword'は読み込めていることを確認
- よってSQL文に問題がある可能性が高い

tkat0 replied:
> よってSQL文に問題がある可能性が高い

これは正しそうです 👍 

変数の埋め込み方を調べてみてください。
オフィシャルドキュメントだとこのあたりです。
https://docs.python.org/ja/3.10/library/sqlite3.html#sqlite3-placeholders

私も調べてみて、すこしわかりにくいなと思った（上記のドキュメントだけでは解決できないかもしれません）ので、上記で解決しなければ再度聞いてください 👍

manami-bunbun replied:
ありがとうございます！こちらもORJSONResponseを入れて、ワイルドカードを消したところ正しいレスポンスが返ってきました。ずっと詰まっていたので大変助かりました。ありがとうございます。

manami-bunbun committed:
step3-6

manami-bunbun committed:
step3-6 done

manami-bunbun committed:
step3-7 done

manami-bunbun committed:
Update point.md

tkat0 commented on python/main.py:
```diff
@@ -18,15 +54,94 @@
+def search_item(keyword: str = None):
+    conn = sqlite3.connect('../db/item.db')
+    c = conn.cursor()
+    items = c.execute('SELECT * FROM items WHERE name LIKE  ? ;', (keyword,)).fetchall()
```
もし、もともとトライされていたようにLIKE %keyword%のような部分一致検索がしたいのであれば、このようにするとできるようです。

```python
items = c.execute("SELECT * FROM items WHERE name LIKE ?", (f"%{keyword}%",)).fetchall()
```

manami-bunbun replied:
ありがとうございます！修正しました

tkat0 commented on point.md:
```diff
@@ -0,0 +1,128 @@
+## Step1 : Git
```
学んだことをまとめるのとても良いですね

tkat0 commented on python/main.py:
```diff
@@ -18,15 +54,94 @@
+# curl -X GET 'http://127.0.0.1:9000/items/89'
+@app.get("/items/", response_class=ORJSONResponse)
+def show_detailById(item_id: int):
+    conn = sqlite3.connect('../db/item.db')
```
ここは課題としては`mercari.sqlite3`が読まれることを期待しています。

もしかして、commitされているdb/item.dbはスキーマではなくデータベースファイル(mercari.sqlite3)相当になってないでしょうか？おそらく手違いで書き換えてしまったのでしょう。

step3.ja.mdによると以下なので、時間があるときにでも対応してみてください

>`items.db`はgitの管理対象にしますが、`mercari.sqlite3`はgitの管理対象として追加しないようにしてください。

(なぜsqlite3ファイルはgitの管理対象にすべきでないのでしょうか? :smiley:

manami-bunbun replied:
こちらは、私が.dbと.sqlite3の存在意義がそれぞれ曖昧になっていて、
- items.db -> schema.sql
- mercari.sqlite3 -> item.db
にしたままでした！ご指摘いただきありがとうございます。

tkat0 commented on python/main.py:
```diff
@@ -18,15 +54,94 @@
     logger.info(f"Receive item: {name}")
+    logger.info(f"Receive item: {category}")
+    logger.info(f"Receive item: {image_name}") 
+    id = random.randint(1,30)
```
テーブルのレコードごとに一意に特定できるuniqueなidを作る必要があります。
それを意図していると思いますが、randint(1,30)では1~30までの整数なので、同じidが複数生成されることが想像できます。
そのため、UUIDなどを使うのがベターです

この方の記事が、全体像やキーワードを理解する上で良さそうです。（色々方法があり、状況に応じて使い分けます）
https://zenn.dev/j5ik2o/articles/a085ab3e3d0f197f6559

（今回は、単純にrandint(1,1000)のようにして乗り切っても良いかもしれません

manami-bunbun replied:
リファレンスありがとうございます！
idの生成の方法がいろいろあって面白そうなので、一旦`randint(1,1000)`にして後で戻ってきます！

KengoA replied:
INT型のIDをUniqueに設定するのは、上の記事でも参照されているAUTO_INCREMENTが適していると思います。
以下のような形でTableのSchemaを指定するだけでSQLite3が自動的にIDを振ってくれるようになり、INSERT時はID以外を指定すればよくなるのでとても便利です。
Table作成: https://github.com/KengoA/mercari-build-training-2022/blob/66c2dd4d315b2369b86fa6ccb1f00a6554ce60ec/scripts/setup_sqlite3.sh#L4 

INSERT: https://github.com/KengoA/mercari-build-training-2022/blob/66c2dd4d315b2369b86fa6ccb1f00a6554ce60ec/go/app/main.go#L203

ちなみに、random.randintだけだとIDの衝突はかなり頻繁に起こってしまいます。以下のサンプルコードを何度か実行してみましょう。
```python
import random

item_size = 30
ids = []

for i in range(item_size):
  id = random.randint(0,1000)
  if id in ids:
    print(f'id collision at the {i}th item')
  else:
    ids.append(id)
```

manami-bunbun replied:
randomだとidが衝突するコードの例までありがとうございます。

違うブランチで作業してしまったのですが、step5のブランチの方でSQL部分を直してIDをAUTO_INCREMENTで割り振りました！添付していただいたSQLの例でVARCHAR型を用いられているのも勉強になりました。

今気づいたのですが、デバッグの利便性ためにCREATE TABLEを`main.py`に貼り付けてgitに上げてしまっていますが、SQL injection等のセキュリティの観点から、gitにはあげない方が良いですよね。後で直してpushします

tkat0 reviewed (approved):
LGTM

お疲れさまでした。気づいた点をコメントしました。
すぐ直す必要はありません。次に進んで、もし時間があまったらトライしてみてください 👍

manami-bunbun committed:
minor change for step3

manami-bunbun committed:
edit in md
Merge branch 'step3' of https://github.com/manami-bunbun/mercari-build-training-2022 into step3
